### PR TITLE
DT_FLASH_AREA replacement with new dts macros

### PR DESCRIPTION
**Pull request by galak**

### CODE REVIEW

**Comment by galak** [comment]
This is based off of https://github.com/zephyrproject-rtos/zephyr/pull/24522.

There's a known issue with samples/subsys/fs/littlefs on nrf52840dk_nrf52840 failing.  This is because the nrf52840dk_nrf52840 has 2 fixed-partition nodes.  We need to decide if partitions should be seen as per flash device or more global.

**Comment by pabigot** [comment]
> There's a known issue with samples/subsys/fs/littlefs on nrf52840dk_nrf52840 failing.

I wasn't aware of that issue.  That's a very ugly failure message.

> This is because the nrf52840dk_nrf52840 has 2 fixed-partition nodes. We need to decide if partitions should be seen as per flash device or more global.

Aha.  Now I finally see what may be underlying the discussion at https://github.com/zephyrproject-rtos/zephyr/pull/24522#discussion_r416022811

Flash partitions identified by `label` property should be global; i.e. the name should be sufficient to identify the containing device and region within the device.  If the numeric ID is local to the containing device it shouldn't matter.

**Comment by de-nordic** [comment]
> 
> 
> > There's a known issue with samples/subsys/fs/littlefs on nrf52840dk_nrf52840 failing.

I have seen that in original PR for FLASH_AREA. I have posted workaround PR here:
https://github.com/zephyrproject-rtos/zephyr/pull/25062

**Comment by galak** [comment]
This should be fully functional now.

**Comment by galak** [comment]
@de-nordic this should supersede #25062, #24848, #24644, and #24514.  We should pull the tests out of #25062. (but not wanting to add more to this PR to try to get it to closure).

File: scripts/dts/gen_defines.py
Comment (mbolivar-nordic) [inline scripts/dts/gen_defines.py:387]:
```suggestion    if node.children:
        out_dt_define(f"{node.z_path_id}_FOREACH_CHILD(fn)",
                      " ".join(f"fn(DT_{child.z_path_id})" for child in
                               node.children.values()))

```

File: scripts/dts/gen_defines.py
Comment (mbolivar-nordic) [inline scripts/dts/gen_defines.py:256]: Why not just emit `{node.z_path_id}_DEP_ORDINAL` for every node? Less of a special case that way.

Comment (galak) [reply scripts/dts/gen_defines.py:256]: I was being conservative at this point.  I'm good w/renaming the macro to `_DEP_ORDINAL`.  I'd like to limit it to `fixed-partitions` for now until we figure out the next phase of DT w/regards to 'struct device'

File: scripts/dts/gen_defines.py
Comment (mbolivar-nordic) [inline scripts/dts/gen_defines.py:588]: Is this meant as a good-enough-for-now solution? I am asking because I'm not sure labels are unique in general. I know non-unique labels in zephyr devicetrees break device_get_binding(), so we rely on unique labels, but I'm asking since we are getting away from requiring label properties in general.

Is it possible to use a node label like `boot_partition` instead of a label property like `mcuboot` here? Node labels are unique, after all.

File: scripts/dts/gen_defines.py
Comment (mbolivar-nordic) [reply scripts/dts/gen_defines.py:256]: Okay. Reasonable.

File: scripts/dts/gen_defines.py
Comment (mbolivar-nordic) [reply scripts/dts/gen_defines.py:588]: I think I have answered my own question. I am OK with labels since dtspec says the value of the property is binding-specific and upstream makes it seem like they're expected to seem unique, as pointed out by @pabigot earlier.

File: Kconfig.zephyr
Comment (mbolivar-nordic) [inline Kconfig.zephyr:353]: Can just remove the line. This will probably anger the linter.

File: scripts/dts/gen_defines.py
Comment (galak) [reply scripts/dts/gen_defines.py:588]: Its meant as good-enough and maintaining the functionality we have today.  I'd prefer to stay away from nodelabel's here especially with the long standing use of the `label` property.

I'd be interested to see how we progress the flash map API w/regards to multiple flash's.

We can also add something to the binding (upstream) about label being unique.

File: subsys/dfu/img_util/flash_img.c
Comment (mbolivar-nordic) [inline subsys/dfu/img_util/flash_img.c:20]: stale

File: subsys/dfu/boot/mcuboot.c
Comment (mbolivar-nordic) [inline subsys/dfu/boot/mcuboot.c:75]: also stale here

File: Kconfig.zephyr
Comment (mbolivar-nordic) [reply Kconfig.zephyr:353]: Doesn't this break mcuboot?

```
=== running "rg DT_FLASH_AREA" in mcuboot (bootloader/mcuboot):
scripts/assemble.py
37:offset_re = re.compile(r"^#define DT_FLASH_AREA_([0-9A-Z_]+)_OFFSET(_0)?\s+(0x[0-9a-fA-F]+|[0-9]+)$")
38:size_re   = re.compile(r"^#define DT_FLASH_AREA_([0-9A-Z_]+)_SIZE(_0)?\s+(0x[0-9a-fA-F]+|[0-9]+)$")
boot/zephyr/include/target.h
38:    !defined(DT_FLASH_AREA_IMAGE_0_OFFSET) || \
39:    !defined(DT_FLASH_AREA_IMAGE_0_SIZE) || \
40:    !defined(DT_FLASH_AREA_IMAGE_1_OFFSET) || \
41:    !defined(DT_FLASH_AREA_IMAGE_1_SIZE) || \
42:    (!defined(CONFIG_BOOT_SWAP_USING_MOVE) && !defined(DT_FLASH_AREA_IMAGE_SCRATCH_OFFSET)) || \
43:    (!defined(CONFIG_BOOT_SWAP_USING_MOVE) && !defined(DT_FLASH_AREA_IMAGE_SCRATCH_SIZE))
sim/mcuboot-sys/csupport/devicetree.h
12:#define DT_FLASH_AREA_IMAGE_0_ID 1
13:#define DT_FLASH_AREA_IMAGE_1_ID 2
14:#define DT_FLASH_AREA_IMAGE_SCRATCH_ID 3
15:#define DT_FLASH_AREA_IMAGE_2_ID 4
16:#define DT_FLASH_AREA_IMAGE_3_ID 5
boot/zephyr/include/sysflash/sysflash.h
16:                                         DT_FLASH_AREA_IMAGE_0_ID : \
17:                                         DT_FLASH_AREA_IMAGE_0_ID)
19:                                         DT_FLASH_AREA_IMAGE_1_ID : \
20:                                         DT_FLASH_AREA_IMAGE_1_ID)
27:                                         DT_FLASH_AREA_IMAGE_0_ID : \
29:                                         DT_FLASH_AREA_IMAGE_2_ID : \
32:                                         DT_FLASH_AREA_IMAGE_1_ID : \
34:                                         DT_FLASH_AREA_IMAGE_3_ID : \
41:#define FLASH_AREA_IMAGE_SCRATCH    DT_FLASH_AREA_IMAGE_SCRATCH_ID
```

**Review by mbolivar-nordic — APPROVED** [review]
Looks overall reasonable as long as the generated IDs are consistent with the legacy IDs and mcuboot is up to date.

File: subsys/dfu/img_util/flash_img.c
Comment (galak) [reply subsys/dfu/img_util/flash_img.c:20]: Fixed

File: subsys/dfu/boot/mcuboot.c
Comment (galak) [reply subsys/dfu/boot/mcuboot.c:75]: Fixed

File: include/devicetree/fixed-partitions.h
Comment (galak) [reply include/devicetree/fixed-partitions.h:36]: Updated the comments, let me know if we still need to clarify more.

File: Kconfig.zephyr
Comment (galak) [reply Kconfig.zephyr]: Fixed mcuboot.

**Review by pabigot — DISMISSED** [review]
Bisect shows

> edtlib: handle match of fixed-partition on any bus

causes breakage in `samples/subsys/fs/littlefs` that is reproduced on head when `CONFIG_LEGACY_DEVICETREE_MACROS=y`:
```
In file included from ../../../../../include/devicetree.h:31,
                 from ../../../../../include/arch/arm/aarch32/arch.h:20,
                 from ../../../../../include/arch/cpu.h:19,
                 from ../../../../../include/kernel_includes.h:39,
                 from ../../../../../include/kernel.h:17,
                 from /mnt/nordic/zp/zephyr/arch/arm/core/offsets/offsets_aarch32.c:25,
                 from /mnt/nordic/zp/zephyr/arch/arm/core/offsets/offsets.c:12:
zephyr/include/generated/devicetree_legacy_unfixed.h:2612: warning: "DT_FIXED_PARTITIONS_0_SIZE" redefined
 2612 | #define DT_FIXED_PARTITIONS_0_SIZE                  65536
      | 
zephyr/include/generated/devicetree_legacy_unfixed.h:2239: note: this is the location of the previous definition
 2239 | #define DT_FIXED_PARTITIONS_0_SIZE                  49152
      | 
zephyr/include/generated/devicetree_legacy_unfixed.h:2614: warning: "DT_FIXED_PARTITIONS_0_LABEL" redefined
 2614 | #define DT_FIXED_PARTITIONS_0_LABEL                 "storage"
      | 
zephyr/include/generated/devicetree_legacy_unfixed.h:2241: note: this is the location of the previous definition
 2241 | #define DT_FIXED_PARTITIONS_0_LABEL                 "mcuboot"
      | 
```

because it has:
```
#define DT_FIXED_PARTITIONS_0_BASE_ADDRESS          0x0
#define DT_FIXED_PARTITIONS_0_SIZE                  49152
#define DT_FIXED_PARTITIONS_0_LABEL                 "mcuboot"
#define DT_FIXED_PARTITIONS_0_READ_ONLY             0
#define DT_FIXED_PARTITIONS_0_BASE_ADDRESS          0x0
#define DT_FIXED_PARTITIONS_0_SIZE                  65536
#define DT_FIXED_PARTITIONS_0_LABEL                 "storage"
#define DT_FIXED_PARTITIONS_0_READ_ONLY             0
```
while the original produced only a single setting (the one for `mcuboot`).

**Comment by galak** [comment]
> Bisect shows
> 
> > edtlib: handle match of fixed-partition on any bus
> 
> causes breakage in `samples/subsys/fs/littlefs` that is reproduced on head when `CONFIG_LEGACY_DEVICETREE_MACROS=y`:
> 
> ```
> In file included from ../../../../../include/devicetree.h:31,
>                  from ../../../../../include/arch/arm/aarch32/arch.h:20,
>                  from ../../../../../include/arch/cpu.h:19,
>                  from ../../../../../include/kernel_includes.h:39,
>                  from ../../../../../include/kernel.h:17,
>                  from /mnt/nordic/zp/zephyr/arch/arm/core/offsets/offsets_aarch32.c:25,
>                  from /mnt/nordic/zp/zephyr/arch/arm/core/offsets/offsets.c:12:
> zephyr/include/generated/devicetree_legacy_unfixed.h:2612: warning: "DT_FIXED_PARTITIONS_0_SIZE" redefined
>  2612 | #define DT_FIXED_PARTITIONS_0_SIZE                  65536
>       | 
> zephyr/include/generated/devicetree_legacy_unfixed.h:2239: note: this is the location of the previous definition
>  2239 | #define DT_FIXED_PARTITIONS_0_SIZE                  49152
>       | 
> zephyr/include/generated/devicetree_legacy_unfixed.h:2614: warning: "DT_FIXED_PARTITIONS_0_LABEL" redefined
>  2614 | #define DT_FIXED_PARTITIONS_0_LABEL                 "storage"
>       | 
> zephyr/include/generated/devicetree_legacy_unfixed.h:2241: note: this is the location of the previous definition
>  2241 | #define DT_FIXED_PARTITIONS_0_LABEL                 "mcuboot"
>       | 
> ```
> 
> because it has:
> 
> ```
> #define DT_FIXED_PARTITIONS_0_BASE_ADDRESS          0x0
> #define DT_FIXED_PARTITIONS_0_SIZE                  49152
> #define DT_FIXED_PARTITIONS_0_LABEL                 "mcuboot"
> #define DT_FIXED_PARTITIONS_0_READ_ONLY             0
> #define DT_FIXED_PARTITIONS_0_BASE_ADDRESS          0x0
> #define DT_FIXED_PARTITIONS_0_SIZE                  65536
> #define DT_FIXED_PARTITIONS_0_LABEL                 "storage"
> #define DT_FIXED_PARTITIONS_0_READ_ONLY             0
> ```
> 
> while the original produced only a single setting (the one for `mcuboot`).

What should be done here, we fixed a bug w/fixed-partitiions on SPI devices in EDT lib.  This exposes an issue in the old generator.  We can add a feature option to EDT to maintain the old behavior for the old generator.

**Comment by pabigot** [comment]
> What should be done here, we fixed a bug w/fixed-partitiions on SPI devices in EDT lib. This exposes an issue in the old generator. We can add a feature option to EDT to maintain the old behavior for the old generator.

I don't know; I'm just  concerned that we won't have tested whether `CONFIG_LEGACY_DEVICETREE_MACROS=y` still works after this and #25009 go in.

**Comment by galak** [comment]
> > 
> 
> I don't know; I'm just concerned that we won't have tested whether `CONFIG_LEGACY_DEVICETREE_MACROS=y` still works after this and #25009 go in.

I think that would be the case regardless of CONFIG_LEGACY_DEVICETREE_MACROS setting as we'd mark the old macros as __deprecated otherwise and never have any in tree references/users.  I guess we'd see this particular issue of duplicated defines, but I'm sure there are other issues we'd miss.

**Comment by mbolivar-nordic** [comment]
Test results:

assemble.py runs without error (cc @brix):

```
$ ./bootloader/mcuboot/scripts/assemble.py -b build_mcuboot/ -p build/zephyr/zephyr.signed.bin -s build/zephyr/zephyr.signed.bin -o /tmp/out
partition MCUBOOT, pos=0, offset=0
partition IMAGE_0, pos=37592, offset=49152
partition IMAGE_1, pos=229972, offset=471040
```

smp_svr: I can build and mostly flash with the instructions online in the Zephyr docs (which use west instead of assemble.py), booting the application via mcuboot.

I could not run the DFU upload step, but I think that's because my laptop's Bluetooth stopped working. I see some HCI errors in `dmesg` now and I can't get `list` to work in `bluetoothctl`. So I don't consider this a real failure.

I believe the legacy and new flash area IDs are consistent with one another:

```
$ grep -e PARTITION_ID -e FLASH_AREA_[0-9].*_ID devicetree_*unfixed.h
devicetree_legacy_unfixed.h:#define DT_FLASH_AREA_0_ID                          0
devicetree_legacy_unfixed.h:#define DT_FLASH_AREA_1_ID                          1
devicetree_legacy_unfixed.h:#define DT_FLASH_AREA_2_ID                          2
devicetree_legacy_unfixed.h:#define DT_FLASH_AREA_3_ID                          3
devicetree_legacy_unfixed.h:#define DT_FLASH_AREA_4_ID                          4
devicetree_unfixed.h:#define DT_N_S_soc_S_flash_controller_4001e000_S_flash_0_S_partitions_S_partition_0_PARTITION_ID 0
devicetree_unfixed.h:#define DT_N_S_soc_S_flash_controller_4001e000_S_flash_0_S_partitions_S_partition_c000_PARTITION_ID 1
devicetree_unfixed.h:#define DT_N_S_soc_S_flash_controller_4001e000_S_flash_0_S_partitions_S_partition_73000_PARTITION_ID 2
devicetree_unfixed.h:#define DT_N_S_soc_S_flash_controller_4001e000_S_flash_0_S_partitions_S_partition_da000_PARTITION_ID 3
devicetree_unfixed.h:#define DT_N_S_soc_S_flash_controller_4001e000_S_flash_0_S_partitions_S_partition_f8000_PARTITION_ID 4
```

I can't find anything wrong with this.

**Comment by galak** [comment]
> > What should be done here, we fixed a bug w/fixed-partitiions on SPI devices in EDT lib. This exposes an issue in the old generator. We can add a feature option to EDT to maintain the old behavior for the old generator.
> 
> I don't know; I'm just concerned that we won't have tested whether `CONFIG_LEGACY_DEVICETREE_MACROS=y` still works after this and #25009 go in.

Fair, but isn't that true of any deprecated functionality in Zephyr.  Since any tests would cause warnings they would never pass sanitycheck.

**Comment by pabigot** [comment]
> > I don't know; I'm just concerned that we won't have tested whether `CONFIG_LEGACY_DEVICETREE_MACROS=y` still works after this and #25009 go in.
> 
> Fair, but isn't that true of any deprecated functionality in Zephyr. Since any tests would cause warnings they would never pass sanitycheck.

I'm not talking about sanitycheck, I'm talking about peoples out-of-tree applications where they aren't quite ready to switch to the non-legacy macros.  If there were no expectation that `CONFIG_LEGACY_DEVICETREE_MACROS=y` would work in that case, why would we even keep the option?

**Comment by galak** [comment]
> > > I don't know; I'm just concerned that we won't have tested whether `CONFIG_LEGACY_DEVICETREE_MACROS=y` still works after this and #25009 go in.
> > 
> > 
> > Fair, but isn't that true of any deprecated functionality in Zephyr. Since any tests would cause warnings they would never pass sanitycheck.
> 
> I'm not talking about sanitycheck, I'm talking about people's out-of-tree applications where they aren't quite ready to switch to the non-legacy macros. If there were no expectation that `CONFIG_LEGACY_DEVICETREE_MACROS=y` would work in that case, why would we even keep the option?

The option is meant as the means to "deprecate" the functionality vs marking every define as __deprecated which has some issues if a defined is used in an #ifdef clause.

I'm not disagreeing with your point, I'm just saying we have this issue more broadly than just here for deprecated functionality.

**Comment by galak** [comment]
> > > > I don't know; I'm just concerned that we won't have tested whether `CONFIG_LEGACY_DEVICETREE_MACROS=y` still works after this and #25009 go in.
> > > 
> > > 
> > > Fair, but isn't that true of any deprecated functionality in Zephyr. Since any tests would cause warnings they would never pass sanitycheck.
> > 
> > 
> > I'm not talking about sanitycheck, I'm talking about people's out-of-tree applications where they aren't quite ready to switch to the non-legacy macros. If there were no expectation that `CONFIG_LEGACY_DEVICETREE_MACROS=y` would work in that case, why would we even keep the option?
> 
> The option is meant as the means to "deprecate" the functionality vs marking every define as __deprecated which has some issues if a defined is used in an #ifdef clause.
> 
> I'm not disagreeing with your point, I'm just saying we have this issue more broadly than just here for deprecated functionality.

We probably should have some tests/deprecated/FOO or something that handles maintaining to the the deprecated functionality for the period of time that its still in tree.

**Comment by galak** [comment]
@pabigot I added a tests/deprecated/dts that at least enables `CONFIG_LEGACY_DEVICETREE_MACROS=y` and does some simple flash map tests.

**Comment by galak** [comment]
Is based on needs #25009

**Review by mbolivar-nordic — APPROVED** [review]
This looks to me like it gets the job done of deprecating the legacy devicetree macros for v2.3, which @carlescufi has said is a priority (not shipping with any deprecated functionality enabled, that is). I also don't see anything obvious that it breaks in terms of legacy support.

I think getting @nvlsianpu 's review would be good before merging, though, since he knows this area much better than I do.

As discussed on dev-review last week and Slack this morning, it seems like a few TSC members agree that extending the release freeze by a day or two due to late-breaking DT changes would not be the end of the world.

**Review by nvlsianpu — COMMENTED** [review]
this umbrella PR looks good, although can't accept mcuboot path without deeper analyze I can't do this weekend. Will do as very first thing on Monday.

File: scripts/dts/edtlib.py
Comment (pabigot) [inline scripts/dts/edtlib.py:191]: Please rename the property to either `_fixed_partitions_no_bus` or `_fixed_partitions_on_bus`, whichever is correct.  Either way `partions` isn't right.

**Comment by carlescufi** [comment]
@galak @mbolivar-nordic 
This can be considered a fix since the majority of the API changes have been introduced already, so I propose we merge this after the `-rc1` tag, given that @nvlsianpu cannot review before Monday.

File: include/devicetree/fixed-partitions.h
Comment (pabigot) [inline include/devicetree/fixed-partitions.h:26]: I would like the documentation to be more clear that this is the `label` property in the node, not a node label attached to the node.

Perhaps: 
* `@brief Get fixed partition with "label" property that matches label`
* `@param label lowercase-and-underscores value for label property`

Applies to both macros in this patch.

File: doc/guides/dts/macros.bnf
Comment (pabigot) [inline doc/guides/dts/macros.bnf:153]: Make this `Allows mapping a "label" property to a compatible node.`  Or something that makes clear the label is a property.

**Review by pabigot — CHANGES_REQUESTED** [review]
Some documentation clarifications and one code change.

File: doc/guides/dts/macros.bnf
Comment (galak) [reply doc/guides/dts/macros.bnf:153]: Done

File: include/devicetree/fixed-partitions.h
Comment (galak) [reply include/devicetree/fixed-partitions.h:26]: Update the comments.

File: scripts/dts/edtlib.py
Comment (galak) [reply scripts/dts/edtlib.py:191]: Fixed.

**Review by pabigot — APPROVED** [review]

**Review by pabigot — APPROVED** [review]
Tested with mcuboot and `samples/subsys/mgmt/mcumgr/smp_svr` with #25167 which included filesystem support on an external SPI NOR flash (`storage` moved to external flash in a local overlay not part of that PR).

File: tests/deprecated/dts/CMakeLists.txt
Comment (hakonfam) [inline tests/deprecated/dts/CMakeLists.txt:6]: Change this? Also the commit message for the last commit seems to need some fixing up.

```
This this covers to small aspects of DTS functionality, one that the
legacy generation script continues to generally function and that
FLASH_AREA id's are the same between the old and new generator.
```

**Review by hakonfam — COMMENTED** [review]
just nits, the last three commit messages needs to be double checked

**Review by nvlsianpu — APPROVED** [review]

File: include/storage/flash_map.h
Comment (hakonfam) [inline include/storage/flash_map.h:223]: We should allow users to provide their own definition of these symbols. This could be the default definition included, but (like with `flash_map_default.c`) it should be possible to override this through a kconfig value.

Currently any dowstream with their own partition definitions is required to maintain a patch of this file.

**Review by hakonfam — CHANGES_REQUESTED** [review]

File: include/storage/flash_map.h
Comment (pabigot) [reply include/storage/flash_map.h:223]: My initial thoughts for solving this were to in the future extend this macro to take a second parameter identifying the provider of partition information.  The default behavior would be as if that parameter were `fixed_partition` which would through magic pass to the current solution, but if it were say `partition_mgr` would route to a different macro or possibly ultimately a function.

The thought is that we do need label uniqueness within a given provider of flash region managers.  Attempting to enforce it across all region managers isn't reasonable IMO, and limiting applications to a single region manager (devicetree or other) is also not reasonable.

A way to resolve the conflict is to pair the label with a provider, using the existing devicetree provider as the default.

Comment (nvlsianpu) [reply include/storage/flash_map.h:223]: This was consulted with me. @mbolivar This is the simple way we can inject our partitions info.

Comment (hakonfam) [reply include/storage/flash_map.h:223]: OK, then there must be something like a kconfig option passed in as the second argument to avoid having to patch the file.

I'm fine with this as long as it allows us to not need patches in any of the upstream source files.

**Review by hakonfam — APPROVED** [review]

**Review by nvlsianpu — APPROVED** [review]

File: west.yml
Comment (nvlsianpu) [inline west.yml:89]: upgrade sha

**Review by mbolivar-nordic — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
